### ISSUE TITLE
Getting image urls for each model item on the fly from api by extending BaseGlideUrlLoader but getUrl cause policy violation

### ISSUE THREAD

**Issue by noobdev19** [bumptech/glide#3619]
Trying to get image urls on the fly. By extending BaseGlideUrlLoader. Any network call inside getUrl is causing polity violation.

I have a list of models (about 1000 items) visible on a recyclerview. But I don't have image urls available for them. So I am calling Glide.load(model).... and using model loader to retrieve the image url for that model item on demand. I am doing this by extending BaseGlideUrlLoader and retrieving the image url for that model from an api (does not support batch calls so I can't just load 1000 image urls on demand) inside BaseGlideUrlLoader.getUrl() method. But it's throwing policy violation.

What would be an workaround? A sample code would be very helpful.

**Glide Version**: 4.9.0

**Integration libraries**: None

**Device/Android Version**: Android 8 Phone

**Issue details / Repro steps / Use case background**: 

**Glide load line / `GlideModule` (if any) / list Adapter code (if any)**:
```java
glide.load(model).into(holder.img);
```

**Layout XML**:
```xml
<FrameLayout xmlns:android="...
```

**Stack trace / LogCat**:
```ruby
paste stack trace and/or log here
```

**Comment by sjudd** [comment]
The best way would actually be to obtain the urls entirely outside of Glide ahead of time and cache them.

If you can't do that, you can still use BaseGlideUrlTarget, but instead of doing your network call to get the url in the getUrl method, do it in a DataFetcher that delegates to the url ModelLoader. This is a bit complex to explain, but someone else might be able to elaborate or you might try asking on StackOverflow.

**Comment by noobdev19** [comment]
I did post it there: https://stackoverflow.com/questions/55787002/glide-modelloader-need-to-get-image-url-from-network-on-the-fly/55787409

but unfortunately I haven't found a solution. I went through your Model documentation so I think if you explained I'll be able to understand.

What I want to do is something like 

class MyDataFetcher() {
  constructor(GlideDefaultDataFetcher fetcher);

@Override
  fetch(){
     var url = getImageUrl(model);
     fetcher.fetch(url);
  }
}


Where I store the default glide data fetcher in a variable inside my custom datafetcher and inside by off main thread fetch method, i just supply the default glide fetcher the url. That way I don't have to fiddle/manage the default code in the package. But even after going through your code a few times I am still unable to code this myself. It would be great if you helped me out on this as at this point I just can't seem to get it done.
